src/compile/data/bin.ts

```diff
@@ -16,57 +16,62 @@ function numberFormatExpr(expr: string, format: string) {
   return `format(${expr}, '${format}')`;
 }
 
-function parse(model: Model): Dict<VgTransform[]> {
-  return model.reduceFieldDef(function(binComponent: Dict<VgTransform[]>, fieldDef: FieldDef, channel: Channel) {
-    const bin = model.fieldDef(channel).bin;
-    if (bin) {
-
-      let binTrans: VgTransform = extend({
-        type: 'bin',
-        field: fieldDef.field,
-        as: [field(fieldDef, {binSuffix: 'start'}), field(fieldDef, {binSuffix: 'end'})]
-      },
-        // if bin is an object, load parameter here!
-        typeof bin === 'boolean' ? {} : bin
-      );
-
-      const transform: VgTransform[] = [];
-      if (!binTrans.extent) {
-        const extentSignal = model.getName(fieldDef.field + '_extent');
-        transform.push({
-          type: 'extent',
-          field: fieldDef.field,
-          signal: extentSignal
-        });
-
-        binTrans.extent = {signal: extentSignal};
-      }
+function checkDiscreteDomainOrLegend(model: Model, transform: VgTransform[], fieldDef: FieldDef, channel: Channel) {
+  const hasDiscreteDomainOrHasLegend = hasDiscreteDomain(model.scale(channel).type) || model.legend(channel);
+  if (hasDiscreteDomainOrHasLegend) {
```

Comment:
You haven't checked if the range formula has already been added.  

Hint: if it's added, it's always the last item in the array.